### PR TITLE
Table alias

### PR DESCRIPTION
**Pull request by aljazerzen**
Implements #248 , but uses named args instead of assigns. Does not cover JOINs

### CODE REVIEW

**Comment by aljazerzen** [comment]
I've also split some of the tests into smaller functions, because they were failing even after I've run them once and did `insta review`. This was because the first assert in function failed and did not even run the second one.

**Comment by max-sixty** [comment]
Thanks for working this through @aljazerzen .

Let me try and write down why having this as a named arg doesn't seem consistent. I'm happy to discuss more on Discord if that's better:
- Named args are generally arguments supplied to a pre-specified list of parameters
- Here we're using a _value_ supplied as a parameter — i.e. generally `e:employees` would be passing an argument of `employees` to an `e` parameter. I would almost prefer we had `employees:e` if we had to do a named arg.
- I understand that `e: employees` reads less well than `e:employees` — these should bind quite tightly, so `join e: employees side:left` doesn't read well because it's unclear the `e`&`employees` are bound. I'm very open to alternatives to try and improve this — e.g. we could have a lint / auto-format that put parentheses around them, like `join (e: employees) side:left`.
- We could also try and find a completely different syntax, either specific to aliases or generally for assignments like `derive`. (Though it's difficult to use `=` because it's valid SQL).

I hope this is reasonable and I so appreciate your contributions over the past few days. Let me know your thoughts. Thank you.

**Comment by aljazerzen** [comment]
Well, I am also against this current implementation - parsing has to be changed because its inconsistent.

I had written the code before posting the issue :D But the AST changes as agnostic of the parsing so they can wait here until we decide on what syntax to use.

While I was working on this, I tried to have `employees:e`, but its just the opposite of the assign op (which is hard to distinguish from named arg, especially for newcomers. See #53 for my opinions.)

In both cases we break definition of function that "has fixed number of positional required args and pre-defined named optional args with default values". What I did to `from` is add an equivalent to python **kwargs (which is [reverse of dictionary unpacking](https://stackoverflow.com/a/1769478/2619989)). Rules there are:

- all args are named,
- args with defaults must come after args without,
- when calling, named args must come after positional
- if there is an `*args` it will contain a list of all non-named args
- if there is an `**kwargs` it will contain a dict all non-declared named args

My point here is that we non-declared named args are a thing in some languages and not such a big inconsistency.

So my opinion is to:
- combine assign and named param into one production, see #53
- extend function definition to allow passing non-declared named args. This would not make sense for user-defined functions, so it would only apply to pipeline transformations. So this would be only a change in how *we* think of functions.

**Comment by aljazerzen** [comment]
Parse expression and function calls

Internal compiler changes:

1. PEST now also parses function calls, they are not constructed only in materialization 
    - `Item::Function` is now `Item::FuncDef`, similar to `Item::FuncCall`
2. PEST now also parses expressions as a tree, not just terms
    - solves #154 (and its uber-hack)
    - there is no need for `Item::Terms` or `Item:Idents`
    - there is no need for unnesting
3. materialized AST does not contain function definitions 
4. `NamedExpr` (prev. `Assign`) now wraps all expressions from `Transformation` or `List`
5. I've simplified quite a few chunks of code (`coerce_to_list`, `RunFunctions`, transformation parsing). This was possible mostly because of offloaded work to PEST.

Language changes:

- pipe allows multiple line breaks
- (conseq of 1.)  `[a b]` now parses as `List([FuncCall(name: a, args: [Ident(b)])])`,
- (conseq of 1.) functions must have at least one paramater,
- (conseq of 2.) transformations arguments cannot be direct function calls (see example below),
- (conseq of 4.) aliases for tables and derives,

Stylistic changes:

- Type aliases such as `Items` and `Terms` are confusing. It does not take much more time to type `Vec<Item>`, but it is much easier to understand for someone new to the codebase.
- I don't like `to_owned()`, because it hides what it really goes on under the hood: cloning.


### Function call changes

`filter upper country = "USA"` now parses as 
```
Transformation:
  Filter:
    - Terms:
        - Expr:
            - Ident: upper
    - Terms:
        - Ident: country
        - Raw: "="
        - String: USA
```
... which fails, because `filter` takes only one positional argument.

`filter (upper country) = "USA"` *still* parses as:
```
Transformation:
  Filter:
    - Terms:
        - Expr:
            - FuncCall:
                name: upper
                args:
                  - Expr:
                      - Ident: country
                named_args: []
        - Raw: "="
        - String: USA
```

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs:10]: 👍

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs:93]: 👍

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs:97]: What are your thoughts on this as `Item` rather than `Vec<Item>`? It's basically always an `Expr` — does it reduce the nesting to have a `Vec` here?

**Comment by max-sixty** [comment]
This is quite the Herculean effort! I'm looking through now. Thanks a lot for covering so much ground.

File: prql/src/translator.rs
Comment (max-sixty) [inline prql/src/translator.rs:17]: 👍

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs]: Is this the correct name? It's a bit surprising it's `to_list` but doesn't return an `Item::List`?

(I notice there's a few of these)

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs]: This avoids the clone (though there may be easier ways?).

```suggestion
            self.into_iter().next().unwrap()
```

or (with `use crate::utils::*;`)

```suggestion
            self.into_only().unwrap()
```

Or maybe you know better than me — is this optimized away by the compiler just as well as the move?

**Comment by max-sixty** [comment]
Are we sure that the `*` is being handled correctly? This was one of my concerns about adopting the `*`. It looks like this doesn't quite work:

```
echo "from employees | aggregate [c~ count *] | derive [y~ (x * 3)]" | ./target/debug/prql compile
Unknown function x
```

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs:198]: At first glance, this looks like quite an improvement over my existing code...

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs]: What were your thoughts about having everything be a `NamedExpr`, rather than having an `Item::NamedExpr` type? Both could work; I realize lots of things can have aliases, though is it a simpler construction to have _everything_ be potentially aliased?

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs]: This is reasonable; FYI this is somewhat what the `AstFold` is for — letting this sort of operation run over any type (it could be called `AstMap` tbh).

**Review by max-sixty — COMMENTED** [review]
I have to head out for a bit so I'm submitting the comments I had; a couple are quite minor. Overall this looks awesome. I wasn't so positive about `~` over `:` but if you are, then I will acquiesce :).

Obv it's easier to review a larger number of smaller changes; but I understand that sometimes you start working on something and then keep making changes..!

**Comment by max-sixty** [comment]
I do think that parsing functions is probably better than leaving them as `Ident`s. 

One of the reasons I had left them as `Ident`s was that we don't know whether an `Ident` is a function with no arguments. (see note below on this)

(There's also an elegance that a "function with no arguments" had exactly the same semantics as "a variable"; I don't think that changes in this construction.)

File: prql/src/parser.rs
Comment (max-sixty) [inline prql/src/parser.rs]: Changing `Terms` into the actual `FuncCall` is such an improvement; and makes other parts (e.g. all the unnesting) much simpler. Great idea!

**Comment by max-sixty** [comment]
It might not matter relative to the other benefits of this; but I'm not sure functions without args are working:

```
echo "
  func x = 3

  from a
  derive [b ~ x]
  " | ./target/debug/prql compile


SELECT
  *,
  x AS b
FROM
  a
```

---
Edit: you said this above!
> (conseq of 1.) functions must have at least one paramater,

File: prql/src/parser.rs
Comment (max-sixty) [inline prql/src/parser.rs]:
```suggestion
                        .map_err(|_| anyhow!("From does not support inline expression. You can only pass a table name."))?,
```

**Comment by max-sixty** [comment]
After looking through this more, here's my summary of where we're at:
- Overall it's a huge improvement. Even aside from Table aliases, removing the acrobatics around `Terms` makes the code _both_ simpler and more explicit. Much code that I wrote has been replaced by better code. Thank you!
- Before we merge we should:
  - Resolve the `x * 3` issue
  - I don't quite understand why `filter upper country = "USA"` needs to fail. Could we parse the `upper country` as a function call, since it's between the transformation name `filter` and the operator `=`?
  - Add TODOs in areas that we don't address now which you agree we should follow-up on (or add an Issue for any bigger ones)
- After we merge we should:
  - Decide whether functions can have no arg. It's a bit odd that they can't. But it has little practical value; we could address the issue in the future. It's equivalent to a variable (I had hoped before that we could unify "variable" and "function call with no params" into the same thing).
  - Standardize function names — `list` should refer to a `List`; if it's a `Vec` we can call it `items`? And should structs hold `Box<Item>` and generally contain an `Expr`, or should they hold `Vec<Item>`?
  - Address the `~` vs `:` question! I would lean towards the `:`, but it seems like you prefer the `~`. If so, let's live with that for a while and see how it goes. 
  - Decide whether we should parse into `NamedExpr` first and then into `Expr` (if I'm reading the code correctly — that's the current flow for some of the parsing). I'm still thinking through whether we can instead parse in `pest` into one or the other.

I'm happy to help with these follow-ups. I've been trying to harangue wasm recently and hopefully will have an MVP quite soon.

File: prql/src/ast.rs
Comment (aljazerzen) [reply prql/src/ast.rs:97]: On parser level it's always `expr` and never multiple expressions. Expression in now actually, an expression, something that we (or SQL) can evaluate to. So I prefer AST to have "correctly" expressed types, thus function only having one expression it evaluates to.

And also, it is not always `Expr`. I've replaced `IntoUnnested` with `IntoExpr` that converts list of items into `Expr(list)`, except when there is only one item in the list, when it converts just to `list[0]`.

https://github.com/max-sixty/prql/blob/f90b1c15b919b6ec3210e9a533b722721f9d53fe/prql/src/parser.rs#L751-L752

File: prql/src/ast.rs
Comment (aljazerzen) [reply prql/src/ast.rs]: Yes, this is not the best naming. I had actual lists here, but because we always needed `Vec<_>`, i've simplified.

File: prql/src/ast.rs
Comment (aljazerzen) [reply prql/src/ast.rs]: True that :+1:

File: prql/src/ast.rs
Comment (aljazerzen) [reply prql/src/ast.rs]: Well, we do have `Item::NamedExpr` (that contains `struct NamedExpr`).

The functions on the `NamedExpr` are here because of typing: they are only applicable to `NamedExpr`, so it would not make sense to pollute `Item::` namespace (and `match self {  Item::NamedExpr() }` in each function).

File: prql/src/ast.rs
Comment (aljazerzen) [reply prql/src/ast.rs]: Oh, yes this can go - it was here for some other reason and it's not needed any more.

**Comment by aljazerzen** [comment]
> It might not matter relative to the other benefits of this; but I'm not sure functions without args are working:
> 
> ```
> echo "
>   func x = 3
> 
>   from a
>   derive [b ~ x]
>   " | ./target/debug/prql compile
> 
> 
> SELECT
>   *,
>   x AS b
> FROM
>   a
> ```
> 
> Edit: you said this above!
> 
> > (conseq of 1.) functions must have at least one paramater,

*Function definitions work, but function calls don't.

If calls would work, the result here would be 
```
SELECT
  *,
  3 AS b
FROM
  a
```

**Comment by aljazerzen** [comment]
`filter upper country = "USA"` needs to fail, because
```
join employees id
```
has to fail.

We can have only one of:
- function name and its arguments separated by space,
- function & transformation arguments separated by space

So I've changed function and transformation parameters so they cannot contain direct function calls.

Unfortunately, I think this cannot be fixed without changing the syntax (or reverting to evaluating only functions that exist).

**Comment by aljazerzen** [comment]
I had an idea for functions without arguments: you said that this is basically a variable. So we can process it as variable! In `RunFunctions`, we also inspect idents and replace them with function def, if found!

This way there is no need for `count *` and you can just write `count`.

**Comment by aljazerzen** [comment]
> Decide whether we should parse into NamedExpr first and then into Expr (if I'm reading the code correctly — that's the current flow for some of the parsing). I'm still thinking through whether we can instead parse in pest into one or the other.

I believe yes, sometimes (like in transformation arguments) we want named parsed first. And sometimes, we even don't want it (like in parenthesis).

File: prql/src/ast.rs
Comment (aljazerzen) [reply prql/src/ast.rs]: Just now I've changed it back: it is much nicer not to have everything wrapped into `NamedExpr`. It does provide some nice typing, so you have to discard possible name, but AST should not have such useless wrappers.

**Comment by aljazerzen** [comment]
Now I've realized that `NamedArgs` and `NamedExpr` can be merged into one! I've dropped `~` because we can consistently parse everything with `:`.

**Comment by aljazerzen** [comment]
I believe I've addressed everything, so there is no need for:

> Add TODOs in areas that we don't address now which you agree we should follow-up on (or add an Issue for any bigger ones)

Language changes (now):

- pipe allows multiple line breaks
- (conseq of 1.) [a b] now parses as List([FuncCall(name: a, args: [Ident(b)])]),
- (conseq of 1.) functions with zero parameters have their own namespace (`count` and `count first_name` can coexist)
- (conseq of 2.) transformations arguments and function arguments cannot be direct function calls (filter example),
- (conseq of 4.) aliases for tables and derives are now the same thing as named args

File: prql/src/materializer.rs
Comment (max-sixty) [inline prql/src/materializer.rs]: Could we leave the note about how we could improve; or at least "we could improve the implementation with a merge join"?

Comment (aljazerzen) [reply prql/src/materializer.rs]: Sure

File: prql/src/ast.rs
Comment (max-sixty) [inline prql/src/ast.rs]: Should this be a method? e.g. `pub fn unnamed(self) -> Item {`

(possibly `into_unnamed` given it's a move, but no stress)

Comment (max-sixty) [reply prql/src/ast.rs]: I see we also have [`discard_name`](https://github.com/max-sixty/prql/pull/250/commits/69ec7a1d023f5c448b04eb99b029402022d9c6ab#diff-03d9ab37a5c3933dabd1520ee3aacc9f093aad4272ecc9abf7027f148ddeea23R127).

Maybe we don't need this; given `expr` is pub already, and they're equivalent. (but no need to block the merge, we can come back to these later)

Comment (aljazerzen) [reply prql/src/ast.rs]: This is a leftover relic...

File: prql/src/parser.rs
Comment (max-sixty) [inline prql/src/parser.rs:56]: 👍  on both points. Wish this were easier (One thing Ocaml is better at...)

File: prql/src/parser.rs
Comment (max-sixty) [inline prql/src/parser.rs:302]: This is clever!

File: prql/src/prql.pest
Comment (max-sixty) [inline prql/src/prql.pest:102]: 👍

**Comment by max-sixty** [comment]
Shall we merge? The marathon is complete...

**Review by max-sixty — APPROVED** [review]

**Comment by max-sixty** [comment]
Congrats and thank you @aljazerzen !

**Comment by aljazerzen** [comment]
It's been quite a sprint and a test of how much do I remember "Introduction to compilers: week 1" :wink:

**Pull request merged**
